Add ec2-sru script for 20.2

lucasmoura opened the pull request:
This is an initial work done on the ec2 manual test for the 20.2 SRU release. Right now, we have added the testing for the following bugs:

*[1876312](https://bugs.launchpad.net/cloud-init/+bug/1876312)
*[1863943](https://bugs.launchpad.net/cloud-init/+bug/1863943)
*[1866930](https://bugs.launchpad.net/ubuntu/+source/cloud-init/+bug/1866930)

But, besides that, I just want to verify right now if I am going in the right direction for the ec2 specific testing.

blackboxsw commented on manual/ec2-sru-20.2.txt:
```diff
@@ -0,0 +1,222 @@
+    ssh $sshopts $EC2_INST cloud-id
+
+    echo "Per LP:1876323 expect do see correct output for cloud-init analyze"
+    ssh $sshopts $EC2_INST -- cloud-init analyze show | grep "Total Time"
```
This bug is not verifiable on Ubuntu releases (which you are launching via your `launch-ec2 --series $series` on line 174. 

Amazon Linux is a different Centos/Fedora rpm-based distribution so it's different  than when we call launch-ec2 and spin up an Ubuntu image.

Also Ryan Harper already validated this particular bug with an individual test that exercised what we expect amazon linux 2 log configuration to look like (though he used lxcs to test) https://github.com/cloud-init/ubuntu-sru/blob/9129f45eed386d9074ab8ad5f404adf1271dfeba/bugs/lp-1876323.txt

lucasmoura replied:
Makes total sense, I completely missed that this was not affecting ubuntu ec2 instances. I will drop this then

blackboxsw commented on manual/ec2-sru-20.2.txt:
```diff
@@ -0,0 +1,222 @@
+run_describe_command() {
+    dnsname=$1
+    query=$2
+    aws ec2 describe-instances --filter "Name=dns-name,Values=${dnsname}" --query $query | jq .[]
```
Since you are dropping the quotes " in all call-sites, you could just use jq -r to return the raw output, then you don't need the tr -d elsewhere
```suggestion
    aws ec2 describe-instances --filter "Name=dns-name,Values=${dnsname}" --query $query | jq -r .[]
```

blackboxsw commented on manual/ec2-sru-20.2.txt:
```diff
@@ -0,0 +1,222 @@
+}
+
+setup_network_interface() {
+    network_interface_json=$(create_network_interface $subnet_id)
```
might have problems as you haven't set $subnet_id locally

blackboxsw commented:
@lucasmoura one thing I'd also like to see validated is that we see imdsv2 tokens REDACTED in the logs as well per 
https://git.launchpad.net/cloud-init/commit/?id=87cd040e

lucasmoura commented:
@blackboxsw I have updated the script with the fixes and I have also added more tests to it. I will look for more ec2 bugs that could be tested here. Also, if you know about more bugs to be tested, just let me know

lucasmoura commented on manual/ec2-sru-20.2.txt:
```diff
@@ -0,0 +1,316 @@
+    PROPOSED_SCRIPT=setup_dev_proposed.sh
+fi
+
+while [ $# -ne 0 ]; do
+	case "$1" in
+		# any flag pass through.
+		--preserve-instance) PRESERVE_INSTANCE=true;;
+		--) shift; break;;
+	esac
+	shift;
+done
```
This cmdline parsing is already there:

https://github.com/cloud-init/ubuntu-sru/blob/master/sru-templates/manual/ec2-sru#L19

But I will definitely add the other helper functions I created there

lucasmoura commented on manual/ec2-sru-20.2.txt:
```diff
@@ -0,0 +1,316 @@
+    [ -n "$EC2_INST" ] || { echo "Failed to launch ec2 instance!"; exit 1; }
+    sleep 60
+
+    trap "terminate_instance ${EC2_INST}" ERR
```
I think we should do both. If the script does fail, we should also terminate the instances, right ?

lucasmoura commented:
@blackboxsw Done, I have updated the script. I am now working on producing the outputs and then I will add the helper functions on the base ec2 template

blackboxsw commented on manual/ec2-sru-20.2.45.txt:
```diff
@@ -0,0 +1,2956 @@
+Startup finished in 3.347s (kernel) + 1min 818ms (userspace) = 1min 4.165s 
+graphical.target reached after 1min 56ms in userspace
++ ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=ERROR ubuntu@ec2-3-133-105-240.us-east-2.compute.amazonaws.com -- sudo systemd-analyze blame
+         38.382s snapd.seeded.service
```
This is just nasty, and I believe we are aware of the snap seeding costs on eoan. but waiting 38 seconds for boot is really painful.

lucasmoura replied:
Yes, I look at the past version we did an SRU and the increase seems really big:
https://github.com/cloud-init/ubuntu-sru/blob/master/manual/ec2-sru-19.4.33.txt#L1208

:cry:
